src/Simulation/Native/src/simulator/wavefunction.hpp

```diff
@@ -90,6 +91,52 @@ class Wavefunction
 
     void flush() const
     {
+        //logic to reorder
+        Fusion fg = fused_.get_fusedgates();
```

Comment:
You are still creating copies in these three lines (changing the getters to const& allows the callers to choose whether to copy or not, and these callers have chosen to copy). I find the data-model rather confusing: are we copying because we need snapshot of the data? because there is multi-threaded access? for exception safety? for some other reason?